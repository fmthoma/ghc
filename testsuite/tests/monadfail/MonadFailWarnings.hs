module MonadFailWarnings where

import Data.Functor.Identity
import Control.Monad.ST

-- should warn, because the do-block gets a general Monad constraint,
-- but should have MonadFail
general :: Monad m => m a
general = do
    Just x <- undefined
    undefined

-- should warn, because Identity isn't MonadFail
identity :: Identity a
identity = do
    Just x <- undefined
    undefined

-- should NOT warn, because IO is MonadFail
io :: IO a
io = do
    Just x <- undefined
    undefined

-- should warn, because (ST s) is not MonadFail
st :: ST s a
st = do
    Just x <- undefined
    undefined

-- should NOT warn, because matching against newtype
newtype Newtype a = Newtype a
newtypeMatch :: Identity a
newtypeMatch = do
    Newtype x <- undefined
    undefined

-- should NOT warn, because Data has only one constructor
data Data a = Data a
singleConMatch :: Identity a
singleConMatch = do
    Data x <- undefined
    undefined

-- should NOT warn, because patterns always match
test6, test7, test8, test9 :: Monad m => m a
test6 = do x <- undefined
           undefined
test7 = do ~(x:y) <- undefined
           undefined
test8 = do _ <- undefined
           undefined
test9 = do (a,b) <- undefined
           undefined
