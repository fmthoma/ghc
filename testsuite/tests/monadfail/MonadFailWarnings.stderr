
MonadFailWarnings.hs:13:5: warning:
    Could not deduce (MonadFail m)
      arising from the failable pattern ‘Just x’ (this will become an error in GHC 8.*)
    from the context: Monad m
      bound by the type signature for:
                 general :: Monad m => m a
      at MonadFailWarnings.hs:11:12-25
    Possible fix:
      add (MonadFail m) to the context of
        the type signature for:
          general :: Monad m => m a
    In a stmt of a 'do' block: Just x <- undefined
    In the expression:
      do { Just x <- undefined;
           undefined }
    In an equation for ‘general’:
        general
          = do { Just x <- undefined;
                 undefined }

MonadFailWarnings.hs:29:5: warning:
    No instance for (MonadFail Identity)
      arising from the failable pattern ‘Just x’ (this will become an error in GHC 8.*)
    In a stmt of a 'do' block: Just x <- undefined
    In the expression:
      do { Just x <- undefined;
           undefined }
    In an equation for ‘identity’:
        identity
          = do { Just x <- undefined;
                 undefined }

MonadFailWarnings.hs:45:5: warning:
    No instance for (MonadFail (ST s))
      arising from the failable pattern ‘Just x’ (this will become an error in GHC 8.*)
    In a stmt of a 'do' block: Just x <- undefined
    In the expression:
      do { Just x <- undefined;
           undefined }
    In an equation for ‘st’:
        st
          = do { Just x <- undefined;
                 undefined }

MonadFailWarnings.hs:53:5: warning:
    No instance for (MonadFail ((->) r))
      arising from the failable pattern ‘Just x’ (this will become an error in GHC 8.*)
    In a stmt of a 'do' block: Just x <- undefined
    In the expression:
      do { Just x <- undefined;
           undefined }
    In an equation for ‘reader’:
        reader
          = do { Just x <- undefined;
                 undefined }
